### PR TITLE
Enable use of the designer

### LINKED ISSUE #9388: How to use the VS Designer? Always getting errors
### Discussed in https://github.com/gitextensions/gitextensions/discussions/9376

<div type='discussions-op-text'>

<sup>Originally posted by **davidgiga1993** July 14, 2021</sup>
Hi,
I'm trying to edit some GUI items but no matter which control I open in the editor I get the following exception:
```
at GitUI.ThreadHelper.ThrowIfNotOnUIThread(String callerMemberName) in D:\Development\Projects\gitextensions\GitExtUtils\GitUI\ThreadHelper.cs:line 59
at ResourceManager.GitExtensionsControlInitialiser..ctor(GitExtensionsControl control) in D:\Development\Projects\gitextensions\ResourceManager\GitExtensionsControlInitialiser.cs:line 25
at ResourceManager.GitExtensionsControl..ctor() in D:\Development\Projects\gitextensions\ResourceManager\GitExtensionsControl.cs:line 20
at GitUI.GitModuleControl..ctor() in D:\Development\Projects\gitextensions\GitUI\GitModuleControl.cs:line 108  
```
Looks like the condition to check if the code is executed by the designer doesn't work as expected?
Or am I missing something here?

I'm using a fresh VS2019 install with the .net desktop development feature installed</div>

### PR DESCRIPTION
**Pull request by RussKie**
This is likely an ongoing work where we'll need to check every form and/or control we edit and apply `IsDesignMode` check to avoid executing unnecessary code in the Windows Forms designer.

Resolves #9388

### CODE REVIEW

File: GitCommands/Settings/AppSettings.cs [hunk 1]
```diff
@@ -2128,6 +2127,22 @@ private static void SaveEncodings()
...
+            {
+                if (_isDesignMode is null)
+                {
+                    string processName = Process.GetCurrentProcess().ProcessName.ToLowerInvariant();
+                    _isDesignMode = processName.Contains("devenv") || processName.Contains("designtoolsserver");
+                }
+
+                return _isDesignMode.Value;
+            }
+        }
```
Comment (RussKie) [inline GitCommands/Settings/AppSettings.cs:2144]: @KlausLoeffelmann this seems the only viable way to detect the designer I found.

File: ResourceManager/GitExtensionsControlInitialiser.cs [hunk 2]
```diff
@@ -8,38 +10,60 @@ namespace ResourceManager
 {
     internal sealed class GitExtensionsControlInitialiser
     {
+        private static bool? _isDesignMode;
         private readonly ITranslate _translate;
 
-        /// <summary>indicates whether the initialisation has been signalled as complete.</summary>
+        // Indicates whether the initialisation has been signalled as complete.
         private bool _initialiseCompleteCalled;
```
Comment (mstv) [inline ResourceManager/GitExtensionsControlInitialiser.cs:17]: ❔ Shall private members be hidden from xmldoc?

**Review by mstv — APPROVED** [review]
LGTM
Thank you!

**Comment by vbjay** [comment]
Not that I use it but what about https://www.jetbrains.com/rider/
detecting devenv only might be an issue.

**Review by gerhardol — COMMENTED** [review]
Looks OK, have not used it
(but I only accidentally open the designer.)

File: ResourceManager/GitExtensionsControlInitialiser.cs (hunk 2 above)
Comment (RussKie) [reply ResourceManager/GitExtensionsControlInitialiser.cs:17]: Xml-docs are generally applied to public API, as they are used to generate documentation. Annotations of non public API is optional. In this specific case I didn't think it was necessary

**Comment by RussKie** [comment]
> Not that I use it but what about [jetbrains.com/rider](https://www.jetbrains.com/rider/) detecting devenv only might be an issue.

No idea. I don't use Rider, so I can't comment on this.

**Comment by dasiths** [comment]
Sorry I'm late. @RussKie @vbjay  the other way to do this is to have a static that gets during app startup. Which indicates proper loading. All other instances can be treated as designer mode. This will support non VS IDE's as well.

**Pull request merged**